diff --git a/x.c b/x.c
index a2153d7..4061224 100644
--- a/x.c
+++ b/x.c
@@ -84,6 +84,7 @@ typedef struct {
   int w, h;   /* window width and height */
   int ch;     /* char height */
   int cw;     /* char width  */
+  int cyo;    /* char y offset */
   int mode;   /* window state/mode flags */
   int cursor; /* cursor style */
 } TermWindow;
@@ -931,6 +932,7 @@ void xloadfonts(const char *fontstr, double fontsize) {
   /* Setting character width and height. */
   win.cw = ceilf(dc.font.width * cwscale);
   win.ch = ceilf(dc.font.height * chscale);
+  xw.cyo = ceilf(dc.font.height * (chscale - 1) / 2);
 
   FcPatternDel(pattern, FC_SLANT);
   FcPatternAddInteger(pattern, FC_SLANT, FC_SLANT_ITALIC);
@@ -1168,7 +1170,7 @@ int xmakeglyphfontspecs(XftGlyphFontSpec *specs, const Glyph *glyphs, int len,
         font = &dc.bfont;
         frcflags = FRC_BOLD;
       }
-      yp = winy + font->ascent;
+      yp = winy + font->ascent + xw.cyo;
     }
 
     if (mode & ATTR_BOXDRAW) {
@@ -1383,13 +1385,13 @@ void xdrawglyphfontspecs(const XftGlyphFontSpec *specs, Glyph base, int len,
 
   /* Render underline and strikethrough. */
   if (base.mode & ATTR_UNDERLINE) {
-    XftDrawRect(xw.draw, fg, winx, winy + dc.font.ascent * chscale + 1, width,
-                1);
+    XftDrawRect(xw.draw, fg, winx,
+                winy + win.cyo + dc.font.ascent * chscale + 1, width, 1);
   }
 
   if (base.mode & ATTR_STRUCK) {
-    XftDrawRect(xw.draw, fg, winx, winy + 2 * dc.font.ascent * chscale / 3,
-                width, 1);
+    XftDrawRect(xw.draw, fg, winx,
+                winy + win.cyo + 2 * dc.font.ascent * chscale / 3, width, 1);
   }
 
   /* Reset clip to none. */
