diff --git a/config.h b/config.h
index 1fbf057..b52800e 100644
--- a/config.h
+++ b/config.h
@@ -187,23 +187,35 @@ static MouseShortcut mshortcuts[] = {
 };
 
 /* Internal keyboard shortcuts. */
-#define MODKEY Mod1Mask
-#define TERMMOD (ControlMask | ShiftMask)
+#define TERMMOD Mod1Mask
+#define TERMMODS TERMMOD | ShiftMask
+#define SCRIPT Mod3Mask
+
+#define SHCMD(cmd)                                                             \
+  {                                                                            \
+    .v = (const char *[]) { "/bin/sh", "-c", cmd, NULL }                       \
+  }
 
 static Shortcut shortcuts[] = {
     /* mask                 keysym          function        argument */
-    {XK_ANY_MOD, XK_Break, sendbreak, {.i = 0}},
-    {ControlMask, XK_Print, toggleprinter, {.i = 0}},
-    {ShiftMask, XK_Print, printscreen, {.i = 0}},
-    {XK_ANY_MOD, XK_Print, printsel, {.i = 0}},
-    {TERMMOD, XK_Prior, zoom, {.f = +1}},
-    {TERMMOD, XK_Next, zoom, {.f = -1}},
-    {TERMMOD, XK_Home, zoomreset, {.f = 0}},
-    {TERMMOD, XK_C, clipcopy, {.i = 0}},
-    {TERMMOD, XK_V, clippaste, {.i = 0}},
-    {TERMMOD, XK_Y, selpaste, {.i = 0}},
-    {ShiftMask, XK_Insert, selpaste, {.i = 0}},
-    {TERMMOD, XK_Num_Lock, numlock, {.i = 0}},
+
+    // Copy-paste
+    {TERMMOD, XK_y, clipcopy, {.i = 0}},
+    {TERMMOD, XK_p, clippaste, {.i = 0}},
+
+    // Zoom
+    {TERMMODS, XK_H, zoom, {.f = -5}},
+    {TERMMODS, XK_L, zoom, {.f = +5}},
+
+    // Scroll
+    {TERMMOD, XK_j, kscrolldown, {.i = 9}},
+    {TERMMOD, XK_k, kscrollup, {.i = 9}},
+    {TERMMODS, XK_J, kscrolldown, {.i = 18}},
+    {TERMMODS, XK_K, kscrollup, {.i = 18}},
+
+    // Urls
+    {SCRIPT, XK_y, externalpipe, SHCMD("url copy")},
+    {SCRIPT, XK_o, externalpipe, SHCMD("url open")},
 };
 
 /*
